player/loadfile.c

```diff
@@ -1703,6 +1703,11 @@ static void play_current_file(struct MPContext *mpctx)
 
     m_config_restore_backups(mpctx->mconfig);
 
+    if (mpctx->stop_play == AT_END_OF_FILE ||
+        mpctx->stop_play == PT_NEXT_ENTRY ||
+        mpctx->stop_play == PT_CURRENT_ENTRY)
+        mp_clear_playback_resume(mpctx, mpctx->filename);
+
```

Comment:
Instead of doing this, why not call the function in the switch/case statement farther down?

Reply:
The goal of that switch seems to be rather different, with the relevant `stop_play` values overlapping in a disjoint way. I don't see a reasonable way to combine them. Though, I will say that I did not see an "obviously correct" place for where to insert this check within the function, so its position is arbitrary to some extent.

Reply:
Do you need `PT_NEXT_ENTRY` or `PT_CURRENT_ENTRY` though? I thought the goal was to delete it at the end of playback of a file.

Reply:
I believe so. Otherwise, if you navigate to the next file, then re-open mpv with the same playlist, it will play the file that was navigated away from.

Reply:
That behavior is intentional although it can certainly be argued that said behavior could be changed/made optional.

Reply:
Yes. See #8138.

Reply:
The intent of the condition which activates "watch later" deletion in this patch is "I am done watching this file". Activating it **only** at `AT_END_OF_FILE` would not fix much in practice.

An alternative to this patch would be to extend the Lua API to 1) communicate to scripts if playback stopped due to navigating to another file or if the player is exiting, and 2) make `mp_clear_playback_resume` callable from Lua scripts; then, the actual behavior could be implemented in the same script which calls `write-watch-later-config`. And, of course, the most thorough alternative would be to take the bull by the horns and fix #8138. Here I opted for the approach which required the least changes in mpv.

Reply:
Right, but if you want to change that behavior that you should modify how playlists are handled internally Pretty sure just not executing `mp_check_playlist_resume` gets you want you want so a user option could be pretty easily added for this.

As for this change, I think the use case is that you'd periodically save data with `write-watch-later-config` as a safeguard. If the file is actually played to the end, then it gets deleted since the user probably doesn't want to "watch later" (otherwise they would have quit, skipped to the next item in the playlist, etc.) That part makes sense to me.

Reply:
> Right, but if you want to change that behavior that you should modify how playlists are handled internally.

Sorry, I don't understand what you mean by this.

> Pretty sure just not executing `mp_check_playlist_resume` gets you want you want so a user option could be easily added for this.

I don't understand which goal would be furthered by not calling `mp_check_playlist_resume`. The use case is resuming playback of playlists correctly, so, `mp_check_playlist_resume` would be an integral part of that (sans some external mechanism which replaces mpv's playlist playback position saving and resuming logic).

> As for this change, I think the use case is that you'd periodically save data with `write-watch-later-config` as a safeguard. If the file is actually played to the end, then it gets deleted since the user probably doesn't want to "watch later" (otherwise they would have quit, skipped to the next item in the playlist, etc.) That part makes sense to me.

Yes. I tried to describe this in the commit message, perhaps my wording wasn't sufficiently clear. I also included a link to a script which does this and makes use of the changes in this patch.

Reply:
>I don't understand which goal would be furthered by not calling mp_check_playlist_resume.

If you don't want a playlist to automatically start from the first file that has watch later data, but just to resume from the saved position when you actually navigate/reach to that spot. Seems like a valid use case and you mentioned that earlier in this comment thread (guess that doesn't address everything though). As for what you listed in #8138, the watch_later stuff all works on a per-file basis not on a playlist basis hence why you see that behavior.

Reply:
> If you don't want a playlist to automatically start from the first file that has watch later data, but just to resume from the saved position when you actually navigate/reach that file in the playlist. Seems like a valid use case and you mentioned that earlier in this comment thread (guess that doesn't address everything though).

I don't think I did; if something I said implied that, then I misspoke, sorry.

If we want to consider users wanting to detach playlist position and individual file position, then that's a whole other use case. Here and in #8138 I only consider the perspective where there exists one single playback position per playlist, which consists of the file within the playlist, and then the position within the file.

> As for what you listed in #8138, the watch_later stuff all works on a per-file basis not on a playlist basis hence why you see that behavior.

Yes. I linked to the relevant code there as well (`mp_check_playlist_resume` again), so I think we've already established that we are on the same page here.

It seems to me like we're beginning to talk past each other, so perhaps it would be better if you could list any pending questions you have that I can address, without referencing prior discussion.

Reply:
I see what you mean. I admit I haven't really considered this case, as it seems to rely on an artifact of mpv's current playlist playback position model, which I understand is what we have out of simplicity rather than intentional design; but even so, I agree that it may be unfair to dismiss it outright based on that argument alone.

I can't judge if continuing to support this hypothetical use case warrants a more complicated patch, but I'd be willing to have a go at whatever solution we decide is best (though something as comprehensive as fixing #8138 would probably be better tackled by a veteran).

One more possible approach is to add an optional flag for `write-watch-later-config` which tells mpv whether it should delete the file when it's done playing. Though the patch would be small, this seems unattractive in that it solidifies a quirk that's an outcome of the current playlist logic as part of the API.

I think I'll revisit the `delete-watch-later-config` command idea, it's starting to look like the best solution after all.

Reply:
Yeah, a `delete-watch-later-config` command sounds like a good idea to me.